tfe/resource_tfe_organization_module_sharing.go

```diff
@@ -0,0 +1,136 @@
+	tfeClient := meta.(*tfe.Client)
+
+	// Get the organization name that will share "produce" modules
+	producer := d.Get("organization").(string)
+	var consumers []string
+	for _, name := range d.Get("module_consumers").([]interface{}) {
+		// ignore empty strings
+		if name == nil {
+			continue
+		}
+		consumers = append(consumers, name.(string))
+	}
+
+	log.Printf("[DEBUG] Create %s module consumers", producer)
+	err := tfeClient.Admin.Organizations.UpdateModuleConsumers(ctx, producer, consumers)
+	if err != nil {
+		return fmt.Errorf("error adding module consumers to %s: %w", producer, err)
+	}
+
+	d.SetId(producer)
+
+	return resourceTFEOrganizationModuleSharingUpdate(d, meta)
+}
+
+func resourceTFEOrganizationModuleSharingUpdate(d *schema.ResourceData, meta interface{}) error {
+	tfeClient := meta.(*tfe.Client)
+
+	// Get the organization name that will share "produce" modules
+	producer := d.Get("organization").(string)
+	var consumers []string
+	for _, name := range d.Get("module_consumers").([]interface{}) {
+		// ignore empty strings
+		if name == nil {
+			continue
+		}
+		consumers = append(consumers, name.(string))
+	}
+
+	log.Printf("[DEBUG] Update %s module consumers", producer)
+	err := tfeClient.Admin.Organizations.UpdateModuleConsumers(ctx, producer, consumers)
+	if err != nil {
+		return fmt.Errorf("error updating module consumers to %s: %w", producer, err)
+	}
+
+	return resourceTFEOrganizationModuleSharingRead(d, meta)
+}
+
+func resourceTFEOrganizationModuleSharingRead(d *schema.ResourceData, meta interface{}) error {
+	tfeClient := meta.(*tfe.Client)
+
+	options := tfe.AdminOrganizationListModuleConsumersOptions{}
+	consumerIDs := make(map[string]string)
+
+	log.Printf("[DEBUG] Read configuration of module sharing for organization: %s", d.Id())
+	for {
+		consumerList, err := tfeClient.Admin.Organizations.ListModuleConsumers(ctx, d.Id(), options)
+		if err != nil {
+			if err == tfe.ErrResourceNotFound {
+				log.Printf("[DEBUG] Organization %s does not longer exist", d.Id())
+				d.SetId("")
+				return nil
+			}
+			return fmt.Errorf("Error reading organization %s mondule consumer list: %w", d.Id(), err)
+		}
+
+		for _, consumer := range consumerList.Items {
+			consumerIDs[consumer.Name] = consumer.ExternalID
+		}
+
+		if consumerList.CurrentPage >= consumerList.TotalPages {
+			break
+		}
+
+		options.PageNumber = consumerList.NextPage
+	}
+
+	d.Set("consumer_ids", consumerIDs)
+
+	return nil
+}
+
+func resourceTFEOrganizationModuleSharingDelete(d *schema.ResourceData, meta interface{}) error {
+	tfeClient := meta.(*tfe.Client)
+
+	log.Printf("[DEBUG] Disable module sharing for organization: %s", d.Id())
+	err := tfeClient.Admin.Organizations.UpdateModuleConsumers(ctx, d.Id(), []string{})
+	if err != nil {
+		if err == tfe.ErrResourceNotFound {
+			return nil
+		}
+		return fmt.Errorf("error deleting organization %s: %w", d.Id(), err)
```

Comment:
this error looks scary you might want to say that we're having trouble deleting the organization module sharing